### PR TITLE
[SPARK-32285][PYTHON] Add PySpark support for nested timestamps with arrow

### PR DESCRIPTION
**Pull request by pralabhkumar**
### What changes were proposed in this pull request?
Added nested timestamp support for Pyspark with Arrow
Following code will run through this PR

```
from pyspark.sql.types import StructType, TimestampType, StructField, ArrayType, LongType
spark.conf.set("spark.sql.execution.arrow.pyspark.fallback.enabled", "False")
spark.conf.set("spark.sql.execution.arrow.pyspark.enabled", "True")

import datetime
import pandas as pd
origin = pd.DataFrame({"a": [[datetime.datetime(2012, 2, 2, 2, 2, 2)]]})
df = spark.createDataFrame(origin, schema=StructType([StructField("a", ArrayType(TimestampType()), True)]))
ts = datetime.datetime(2015, 11, 1, 0, 30)

schema = StructType([StructField("a", ArrayType(TimestampType()), True)])

df = spark.createDataFrame([([ts, ts],)], schema=schema)

df.toPandas()
```

### Why are the changes needed?
This change is required to convert ArrayType(TimeStamp) to pandas via arrow. 

### Does this PR introduce any user-facing change?
Yes user will be able to convert DF which contain Arraytype(Timestamp) to pandas

### How was this patch tested?
unit tests

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
@pralabhkumar mind keeping the GitHub PR template as is? (https://github.com/apache/spark/blob/master/.github/PULL_REQUEST_TEMPLATE)

**Comment by pralabhkumar** [comment]
> @pralabhkumar mind keeping the GitHub PR template as is? (https://github.com/apache/spark/blob/master/.github/PULL_REQUEST_TEMPLATE)

@HyukjinKwon 
Have updated PR template. Sorry since it is still in progress, I missed the PR template .
As stated in Jira , if u can do quick review and let me know if direction is correct , i'll complete the rest of the part of the jira and will improve the PR .

**Comment by pralabhkumar** [comment]
@HyukjinKwon 

Have fixed all the issues with the build . Please look into the PR .

File: python/pyspark/sql/pandas/types.py [hunk 1]
```diff
@@ -20,6 +20,7 @@
 pandas instances during the type conversion.
 """
 
+
```
Comment (HyukjinKwon) [inline python/pyspark/sql/pandas/types.py:23]: Let's remove these newlines.

File: python/pyspark/sql/tests/test_arrow.py [hunk 2]
```diff
@@ -118,32 +117,50 @@ def create_pandas_data_frame(self):
         data_dict["4_float_t"] = np.float32(data_dict["4_float_t"])
         return pd.DataFrame(data=data_dict)
 
-    def test_toPandas_fallback_enabled(self):
+    # def test_toPandas_fallback_enabled(self):
```
Comment (HyukjinKwon) [inline python/pyspark/sql/tests/test_arrow.py:120]: Can you enable the tests back?

File: python/pyspark/sql/pandas/types.py [hunk 3]
```diff
@@ -190,10 +191,19 @@ def _check_series_localize_timestamps(s, timezone):
     """
     from pyspark.sql.pandas.utils import require_minimum_pandas_version
     require_minimum_pandas_version()
-
+    from pandas import Series, DatetimeTZDtype
+    from pandas.api.types import is_datetime64_dtype
     from pandas.api.types import is_datetime64tz_dtype
     tz = timezone or _get_local_timezone()
     # TODO: handle nested timestamps, such as ArrayType(TimestampType())?
+    # Check if its Array of TimeStamp
```
Comment (HyukjinKwon) [inline python/pyspark/sql/pandas/types.py:199]: typo

File: python/pyspark/sql/pandas/types.py [hunk 4]
```diff
@@ -190,10 +191,19 @@ def _check_series_localize_timestamps(s, timezone):
     """
     from pyspark.sql.pandas.utils import require_minimum_pandas_version
     require_minimum_pandas_version()
-
+    from pandas import Series, DatetimeTZDtype
+    from pandas.api.types import is_datetime64_dtype
     from pandas.api.types import is_datetime64tz_dtype
     tz = timezone or _get_local_timezone()
     # TODO: handle nested timestamps, such as ArrayType(TimestampType())?
```
Comment (HyukjinKwon) [inline python/pyspark/sql/pandas/types.py:198]: remove this

File: python/pyspark/sql/pandas/types.py [hunk 5]
```diff
@@ -190,10 +191,19 @@ def _check_series_localize_timestamps(s, timezone):
...
+    # Check if its Array of TimeStamp
+    if datatype == ArrayType.__name__:
```
Comment (HyukjinKwon) [inline python/pyspark/sql/pandas/types.py:200]: does it work for nested + nested array of timestamps?

File: python/pyspark/sql/tests/test_arrow.py [hunk 6]
```diff
@@ -513,12 +531,12 @@ def run_test(num_records, num_parts, max_records, use_delay=False):
...
+            (1024, 512, 2),  # Use large num partitions for more likely collecting out of order
             (64, 8, 2, True),  # Use delay in first partition to force collecting out of order
-            (64, 64, 1),       # Test single batch per partition
-            (64, 1, 64),       # Test single partition, single batch
-            (64, 1, 8),        # Test single partition, multiple batches
-            (30, 7, 2),        # Test different sized partitions
+            (64, 64, 1),  # Test single batch per partition
+            (64, 1, 64),  # Test single partition, single batch
+            (64, 1, 8),  # Test single partition, multiple batches
+            (30, 7, 2),  # Test different sized partitions
```
Comment (HyukjinKwon) [inline python/pyspark/sql/tests/test_arrow.py:539]: let's remove all these unrelated changes.

File: python/pyspark/sql/pandas/types.py (hunk 1 above)
Comment (pralabhkumar) [reply python/pyspark/sql/pandas/types.py]: done

File: python/pyspark/sql/tests/test_arrow.py (hunk 2 above)
Comment (pralabhkumar) [reply python/pyspark/sql/tests/test_arrow.py]: done. However i have changed the expected failure type to Map<TimeStamp,TimeStamp> instead of ArrayType<TimeStamp> . Since ArrayType<TimeStamp> will not throw exception

File: python/pyspark/sql/pandas/types.py (hunk 3 above)
Comment (pralabhkumar) [reply python/pyspark/sql/pandas/types.py]: done

File: python/pyspark/sql/pandas/types.py (hunk 4 above)
Comment (pralabhkumar) [reply python/pyspark/sql/pandas/types.py]: done

File: python/pyspark/sql/pandas/types.py (hunk 5 above)
Comment (pralabhkumar) [reply python/pyspark/sql/pandas/types.py:196]: Currently only one level of nesting is handled . So only ArrayType<TimeStamp> will work . Also this is what being suggested in Jira (as the immediate need)

File: python/pyspark/sql/tests/test_arrow.py (hunk 6 above)
Comment (pralabhkumar) [reply python/pyspark/sql/tests/test_arrow.py]: done

**Comment by pralabhkumar** [comment]
@HyukjinKwon . Thx for the review , have made changes and fix some bugs (Also added more description in PR)

Please review the PR .

**Comment by HyukjinKwon** [comment]
cc @BryanCutler FYI

File: python/pyspark/sql/pandas/types.py [hunk 7]
```diff
@@ -296,7 +337,34 @@ def _check_series_convert_timestamps_localize(s, from_timezone, to_timezone):
...
+    :param series: Pandas series
+    :param to_tz: to timezone
+    :param from_tz: from time zone
+    :return: return series respecting timezone
+    """
+    from pandas.api.types import is_datetime64tz_dtype, is_datetime64_dtype
+    import pandas as pd
+    from pandas import Series
+    data_after_conversion = []
+    for data in series:
```
Comment (BryanCutler) [inline python/pyspark/sql/pandas/types.py:352]: So this is iterating over each timestamp array in the series, applying conversions each time and then building a new series back from a list? That seems pretty inefficient and looks to be a lot of specialized conversion going on here for just arrays of timestamps.

**Review by BryanCutler — COMMENTED** [review]
Thanks for taking this on @pralabhkumar , I think it's fine to only tackle timestamps with one level of nesting but there is also MapType and StructType that could have timestamps, does this check for those?

Also, it looks like the Series of timestamp arrays is iterated with a for loop, which will be inefficient. As an alternative, it might be possible to flatten the Arrow array and apply the existing timestamp conversion only on non-nested arrays. That would probably be more efficient, but I'm not sure how difficult it would be do, so I'm not opposed to the way it's done here.

File: python/pyspark/sql/pandas/types.py [hunk 8]
```diff
@@ -253,14 +278,29 @@ def _check_series_convert_timestamps_internal(s, timezone):
...
+def __modified_series(data, is_array):
+    """
+    :param data: Converted data
+    :param is_array: If the input data type is type of array
+    :return: If input type is array ,then return series with array of data
+    else return series as it is.
+    """
+    from pandas import Series
+    if is_array:
+        return Series([data])
```
Comment (BryanCutler) [inline python/pyspark/sql/pandas/types.py:299]: Sorry, I don't quite understand what this function is doing, it looks like it's making a Series with 1 array?

File: python/pyspark/sql/pandas/types.py (hunk 7 above)
Comment (pralabhkumar) [reply python/pyspark/sql/pandas/types.py]: @BryanCutler  , Have done some changes , remove all the unnecessary specialized conversion and also creating series

File: python/pyspark/sql/pandas/types.py (hunk 8 above)
Comment (pralabhkumar) [reply python/pyspark/sql/pandas/types.py]: @BryanCutler  Removed the logic

**Comment by pralabhkumar** [comment]
@BryanCutler . Thx for finding the time to review it and Sorry for replying late. Have done changes to simplify the conversion logic . Also this PR is only for Array of Timestamp . Later on I can work on Map Type and Struct Type as a separate PR.

**Comment by pralabhkumar** [comment]
@BryanCutler @HyukjinKwon.   Please review the PR , and please let me know if changes are needed.

**Comment by pralabhkumar** [comment]
@BryanCutler @HyukjinKwon  Gentle ping. Please review the PR

**Comment by BryanCutler** [comment]
Apologies, I have not had a chance to review again. I will try to do so later this week.

**Comment by pralabhkumar** [comment]
Thx for your response @BryanCutler . Looking forward to your review

File: python/pyspark/sql/pandas/types.py [hunk 9]
```diff
@@ -294,8 +303,10 @@ def _check_series_convert_timestamps_localize(
     from pandas.api.types import is_datetime64tz_dtype, is_datetime64_dtype
     from_tz = from_timezone or _get_local_timezone()
     to_tz = to_timezone or _get_local_timezone()
-    # TODO: handle nested timestamps, such as ArrayType(TimestampType())?
-    if is_datetime64tz_dtype(s.dtype):
+    if datatype == ArrayType.__name__:
```
Comment (BryanCutler) [inline python/pyspark/sql/pandas/types.py:306]: I don't think it's a good way to check for data types by using the `__name__` attribute, isn't there another way to check that the data type is an ArrayType?

File: python/pyspark/sql/pandas/types.py [hunk 10]
```diff
@@ -356,7 +389,27 @@ def _convert_map_items_to_dict(s: "PandasSeriesLike") -> "PandasSeriesLike":
...
+    checks whether the series contain Timstamp object
+    :param s: pd.series
+    :return:  True if the series contain timestamp object
+    """
+    from numpy import ndarray
+    import datetime
+    from pandas.api.types import is_datetime64tz_dtype, is_datetime64_dtype
+    if not s.empty and isinstance(s.dtype, object) and \
+            isinstance(s.iloc[0], list) and len(s.iloc[0]) > 0 and \
+            isinstance(s.iloc[0][0], datetime.datetime):
```
Comment (BryanCutler) [inline python/pyspark/sql/pandas/types.py:403]: I don't think you can rely on checking the first value to determine the data type

File: python/pyspark/sql/tests/test_arrow.py [hunk 11]
```diff
@@ -133,18 +133,29 @@ def test_toPandas_fallback_enabled(self):
...
             with self.warnings_lock:
                 with self.assertRaisesRegex(Exception, 'Unsupported type'):
                     df.toPandas()
 
+    def test_toPandas_array_timestamp(self):
+        schema = StructType([
+            StructField("idx", LongType(), True),
+            StructField("timestamp_array", ArrayType(TimestampType()), True)])
+        data = [(0, [datetime.datetime(1969, 1, 1, 1, 1, 1)]),
+                (2, [datetime.datetime(2100, 3, 3, 3, 3, 3)])]
```
Comment (BryanCutler) [inline python/pyspark/sql/tests/test_arrow.py:153]: The data here is a little too simple, there should be multiple values including nulls

File: python/pyspark/sql/pandas/types.py [hunk 12]
```diff
@@ -306,8 +317,28 @@ def _check_series_convert_timestamps_localize(
         return s
 
 
+def modify_timestamp_array(data, to_tz: Optional[str], from_tz: Optional[str] = None
+                           , is_utc: bool = False):
+    import pandas as pd
+    if data is None:
+        return [None]
```
Comment (BryanCutler) [inline python/pyspark/sql/pandas/types.py:324]: Wouldn't this change a value in the series from None to an array with one value that is None?

**Review by BryanCutler — COMMENTED** [review]
@pralabhkumar I am currently a -0 on the changes and approach here. I think there are issues with simply checking the first element in the first value to see if it's a timestamp. Also it does not seem like much of the additions here could be used to handle other nested types or deeper nesting, which would then all require specialized functions to handle.

I am thinking it would be better to utilize pyarrow for all type checking and flattening of nested columns so that all conversions are done on flat series. Although, I understand there might be some challenges with this approach as well.

File: python/pyspark/sql/pandas/types.py (hunk 9 above)
Comment (pralabhkumar) [reply python/pyspark/sql/pandas/types.py:306]: Will change it

File: python/pyspark/sql/pandas/types.py (hunk 10 above)
Comment (pralabhkumar) [reply python/pyspark/sql/pandas/types.py:403]: Agree , However for array with pyarrow , I m checking the dtype for the array. 
But I think as suggested by you for pyarrow , it would be better to use pyarrow for type checking

File: python/pyspark/sql/tests/test_arrow.py (hunk 11 above)
Comment (pralabhkumar) [reply python/pyspark/sql/tests/test_arrow.py:153]: Have tested with none . but agree , will add more stringent test cases.

**Comment by pralabhkumar** [comment]
@BryanCutler 
First of all , really thx for your time and effort to review the PR. 
===
 I think there are issues with simply checking the first element in the first value to see if it's a timestamp.
===

Response : I completely agree with you , checking first value is not correct . I'll change it .  

===
Also it does not seem like much of the additions here could be used to handle other nested types or deeper nesting, which would then all require specialized functions to handle
===

Response : modify_timestamp_array is the method which basically does the datatime conversion . We can use the same code for deeper nesting conversion , its calling method have to called it recursively .  However I agree with u that this code may not be utilized for other nested type. 
As per the jira , the immediate requirement was array of timestamp , that's why I only took consideration of Arraytype.  


====
I am thinking it would be better to utilize pyarrow for all type checking and flattening of nested columns so that all conversions are done on flat series. Although, I understand there might be some challenges with this approach as well.
====
Response : I tried to have code similar to what already have written . For e.g do the time conversion , post converting to panda . 
Therefore my approach/design is very same as what already there in code.  

However I would like better understand your approach. If possible may be we can have call to better understand your approach and then I can code the same way.  Please let me know , if you are ok with the same.  If you think current approach is fine , I'll address all the review comments provided by you.

**Comment by pralabhkumar** [comment]
Thx @BryanCutler 

First of all , really thx for your time and effort to review the PR.

########
 I think there are issues with simply checking the first element in the first value to see if it's a timestamp.
########

Response : I completely agree with you , checking first value is not correct . I'll change it .

########
Also it does not seem like much of the additions here could be used to handle other nested types or deeper nesting, which would then all require specialized functions to handle
########

Response : modify_timestamp_array is the method which basically does the datatime conversion . We can use the same code for deeper nesting conversion , its calling method have to called it recursively .  However I agree with u that this code may not be utilized for other nested type. 
As per the jira , the immediate requirement was array of timestamp , that's why I only took consideration of Arraytype.

########
I am thinking it would be better to utilize pyarrow for all type checking and flattening of nested columns so that all conversions are done on flat series. Although, I understand there might be some challenges with this approach as well.
########

Response : I tried to have code similar to what already have written . For e.g do the time conversion , post converting to panda . 
Therefore my approach/design is very same as what already there in code.  

However I would like better understand your approach. If possible may be we can have call to better understand your approach and then I can code the same way.  Please let me know , if you are ok with the same.  

on the other hand if  think current approach is fine , I'll address all the review comments provided by you.
